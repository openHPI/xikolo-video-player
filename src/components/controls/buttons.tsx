import {
  FunctionalComponent,
  h,
} from '@stencil/core';

import * as icon from '../../utils/icon';


interface FullscreenProps {
  fullscreen: boolean;
  onRequest: (e: Event) => void;
  onExit: (e: Event) => void;
}

export const Fullscreen: FunctionalComponent<FullscreenProps> = (props) => {
  if(props.fullscreen) {
    return <button onClick={props.onRequest} innerHTML={icon.Compress} />
  } else {
    return <button onClick={props.onExit} innerHTML={icon.Expand} />
  }
};


interface PlayPauseProps {
  playing: boolean;
  onPlay: (e: Event) => void;
  onPause: (e: Event) => void;
}

export const PlayPause: FunctionalComponent<PlayPauseProps> = (props) => {
  if(!props.playing) {
    return <button onClick={props.onPlay} innerHTML={icon.Play} />
  } else {
    return <button onClick={props.onPause} innerHTML={icon.Pause} />
  }
}
