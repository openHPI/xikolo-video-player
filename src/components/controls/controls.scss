@use "../../utils/variables.scss";

.controls {
  background-color: var(--vp-gutter-color);
  display: flex;
  flex: 1 0 auto;
  flex-flow: row wrap;

  &--fullscreen-mode {
    @media only screen and (min-width: 992px) {
      .controls__subtitle-row {
        font-size: 2.5em;
      }
      .controls__toolbar {
        .controls__playback-rate {
          &__menu {
            // Needed for matching the bottom position of the settings menu.
            margin-bottom: -3px;
            bottom: 60px;
          }

          &__button {
            font-size: 1em;
          }
        }

        .controls__time {
          font-size: 1.2em;
        }

        svg,
        .svg {
          height: 2em;
          width: 2em;
        }
      }
    }
  }

  @media only screen and (min-width: 376px) {
    .controls__toolbar {
      button {
        padding: 0.6em 0.7em;
      }
    }
  }

  @media only screen and (min-width: 992px) {
    .controls__toolbar .controls__slider--volume {
      display: none;
    }
  }

  @media only screen and (max-width: 580px) {
    .controls__subtitle {
      bottom: 40px;

      &-row {
        font-size: 1.3em;
        line-height: 1em;
        padding: 5px;
      }
    }
    .controls__toolbar {
      .controls__slider--volume {
        display: none;
      }

      .controls__time {
        padding: inherit;
      }
    }
  }

  // Reduces the font size for the current time display, so it does not wrap into two lines
  @media only screen and (max-width: 590px) {
    .controls__toolbar {
      .controls__time {
        font-size: 0.75em;
      }
    }
  }
}

.controls__playback-rate {
  position: relative;

  &__menu {
    position: absolute;
    z-index: variables.$z-index-l;
    // Needed for matching the bottom position of the settings menu.
    margin-bottom: -1px;
    bottom: 50px;
    right: 50%;
    transform: translateX(50%);
    background: var(--vp-menu-color);
    font-size: 0.9em;
    line-height: 0.9em;
    padding: 0.2em;
    visibility: hidden;
    opacity: 0;
    transition:
      visibility 0.3s linear,
      opacity 0.3s linear;

    &--open {
      visibility: visible;
      opacity: 1;
    }
  }

  &__button {
    font-size: 0.9em;
    line-height: 0.9em;
    width: 40px;
    box-sizing: content-box;
    text-align: left;
    position: relative;
    text-indent: 1em;

    &:hover {
      background: rgba(255, 255, 255, 0.2);
    }

    &--current {
      font-weight: 600;
      &:before {
        content: "\2022";
        color: var(--vp-main-color);
        position: absolute;
        left: -0.1em;
      }
    }
  }
}

.controls__subtitle {
  position: absolute;
  left: 0px;
  right: 0px;
  z-index: variables.$z-index-l;
  bottom: 60px;
  align-self: center;
  text-align: center;
  max-width: 90%;
  margin: auto;

  &-row {
    background: rgba(50, 50, 50, 0.7);
    padding: 10px;
    margin: 0.2em;
    border-radius: 10px;
    display: inline-block;
    white-space: pre-wrap;
    font-size: 1.5em;
    line-height: 1.3em;
  }
}

.controls__slider {
  flex: 1 1 100%;
  appearance: none;
  overflow: hidden;
  height: var(--vp-control-slider-height);
  background: transparent;
  color: var(--vp-slider-color);
  margin: 0;
  border-top: 1px solid black;
  border-top-color: var(--vp-bg-color);
  border-bottom: 1px solid black;
  border-bottom-color: var(--vp-bg-color);
  cursor: pointer;

  &::-webkit-slider-thumb {
    appearance: none;
    box-shadow: -100vw 0 0 100vw currentColor;
    // Safari need a fixed width to display the slider thumb
    width: 1px;
    height: var(--vp-control-slider-height);
  }

  &::-moz-range-thumb {
    appearance: none;
    box-shadow: -100vw 0 0 100vw currentColor;
    width: 0;
    border: 0;
  }

  &--volume {
    border: 1px solid black;

    &:focus {
      border: 1px solid var(--vp-main-color);
    }
  }
}

.controls__toolbar {
  margin: 1px;
  color: inherit;

  display: flex;
  flex: 1 1 100%;
  flex-flow: row nowrap;
  align-items: center;

  .controls-volume {
    display: flex;
    flex-flow: row nowrap;
    align-items: center;

    &:hover {
      @media only screen and (min-width: 376px) {
        .controls__slider--volume {
          display: block;
        }
      }
    }
  }

  .controls__time {
    flex-grow: 1;
    font-size: 0.9em;
    padding: 0.6em 0.7em;
  }

  .controls__settings-icon {
    display: block;
    overflow: hidden;
    transition-duration: 0.4s;
    transition-property: transform;

    &--open,
    &:hover {
      transform: rotate(90deg);
    }
  }

  .controls__button-icon {
    display: block;
    overflow: hidden;
    transition: box-shadow 0.3s ease-in-out;

    &--active,
    &:hover {
      box-shadow: 0px 2px 1px -1px rgba(231, 166, 26, 0.72);
    }
  }

  button {
    border: 0;
    padding: 0.6em 0.5em;
    margin: 0;
    background: transparent;
    color: inherit;
    cursor: pointer;

    &:hover {
      & .controls__settings-icon {
        transform: rotate(90deg);
      }
      & .controls__subtitle-icon {
        box-shadow: 0px 2px 1px -1px rgba(231, 166, 26, 0.72);
      }
    }
  }

  svg,
  .svg {
    display: block;
    height: 1.5em;
    width: 1.5em;
    fill: currentColor;
    color: inherit;
  }
}
