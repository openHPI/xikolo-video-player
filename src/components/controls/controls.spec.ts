import { newSpecPage } from "@stencil/core/testing";
import { Controls } from '../controls/controls';
import { SettingsMenu } from "../settings-menu/settings-menu";
import { defaultStatus } from '../../utils/status';
import { TextTrack, WebVTT } from '../../utils/webVTT';

describe('xm-controls', () => {
  let page, textTrack: TextTrack, shadowRoot, controls, settingsMenu;

  it('should build', () => {
    expect(new Controls()).toBeTruthy();
    expect(new SettingsMenu()).toBeTruthy();
  });

  beforeEach(async () => {
    page = await newSpecPage({
      html: `<div></div>`,
      components: [Controls, SettingsMenu],
      supportsShadowDom: true
    });

    textTrack = new TextTrack();
    controls = page.doc.createElement('xm-controls');
    controls.status = defaultStatus;
    controls.textTrack = textTrack;
    page.root.appendChild(controls);
    await page.waitForChanges();
    expect(page.rootInstance).toBeTruthy();
    expect(page.root.querySelector('xm-controls')).toBeTruthy();
    shadowRoot = page.root.querySelector('xm-controls').shadowRoot;
    expect(shadowRoot).toBeTruthy();

    // We need to render all child components manually for acces to their shadow doms!
    settingsMenu = shadowRoot.querySelector('xm-settings-menu');
    let settingsInstance = new SettingsMenu();
    settingsInstance.status = defaultStatus;
    settingsInstance.isOpen = false;
    settingsInstance.textTrack = textTrack;
    settingsMenu.innerHTML = settingsInstance.render();
    await page.waitForChanges();
    expect(settingsMenu.shadowRoot).toBeTruthy();
  });

  it('should render the settings-menu', async () => {
    const status = {
      ...controls.status,
      openedSettingsMenu: true,
    };
    expect(settingsMenu.shadowRoot.querySelector('.settings-menu')).toBeTruthy();
    expect(settingsMenu.shadowRoot.querySelector('.settings-menu.menu--open')).not.toBeTruthy();
    expect(shadowRoot.querySelector('.controls__settings-icon')).not.toHaveClass('controls__settings-icon--open');
    let button = shadowRoot.querySelector('.controls__settings-icon').parentNode;
    expect(button).toBeTruthy();
    button.click();
    controls.status = status;
    await page.waitForChanges();
    expect(shadowRoot.querySelector('.controls__settings-icon')).toHaveClass('controls__settings-icon--open');
    expect(settingsMenu.shadowRoot.querySelector('.settings-menu.menu--open')).toBeTruthy();
    expect(page.root).toMatchSnapshot();
  });

  it('should show a list of texttracks', async () =>{
    const de: WebVTT = {
      meta: {
        language: 'de',
        label: 'Deutsch',
      },
      cues: [{
        end: 1,
        identifier: '1',
        start:1,
        styles: '',
        text: 'test',
      }],
      strict: true,
      valid: true,
    };
    const status = {
      ...controls.status,
      subtitle: {
        ...controls.status.subtitle,
        enabled: true,
        language: 'de',
      },
      settings: {
        ...controls.status.settings,
        textTrack: 'de',
      }
    };
    expect(shadowRoot.querySelector('.controls__subtitle-icon')).not.toBeTruthy();
    let textTrackValue = settingsMenu.shadowRoot.querySelector('.settings-menu__button-value').firstChild;
    expect(textTrackValue.nodeValue.trim()).toBe('Off');
    expect(settingsMenu.status.settings.textTrack).toBe('off');
    expect(textTrack.getTextTracks()).toBe(null);
    textTrack.addWebVTT(de);
    await page.waitForChanges();
    expect(textTrack.getTextTracks().length).toBe(1);
    controls.status = status;
    await page.waitForChanges();
    expect(shadowRoot.querySelector('.controls__subtitle-button').querySelector('.controls__shortcut-icon')).toBeTruthy();
    expect(shadowRoot.querySelector('.controls__subtitle-button').querySelector('.controls__shortcut-icon')).toHaveClass('controls__shortcut-icon--active');
    expect(settingsMenu.status.settings.textTrack).toBe('de');
    // must reselect it for testing !
    textTrackValue = settingsMenu.shadowRoot.querySelector('.settings-menu__button-value').firstChild;
    expect(textTrackValue.nodeValue.trim()).toBe('Deutsch');
    const subtitleSubmenubutton = settingsMenu.shadowRoot.querySelector('.settings-menu__button').firstChild;
    subtitleSubmenubutton.click();
    await page.waitForChanges();
    const submenu = settingsMenu.shadowRoot.querySelector('.settings-menu__submenu-content');
    expect(submenu.querySelector('.settings-menu__button')).toBeTruthy();
    expect(submenu.childNodes.length).toBe(2);
    expect(textTrack.getTextTrackValues().length).toBe(2);
    expect(page.root).toMatchSnapshot();
  });

});
