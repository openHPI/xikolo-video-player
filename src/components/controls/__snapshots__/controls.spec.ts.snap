// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`xm-controls should render the settings-menu 1`] = `
<div>
  <xm-controls>
    <mock:shadow-root>
      <div class="controls">
        <xm-settings-menu>
          <mock:shadow-root>
            <div class="menu menu--open settings-menu">
              <div class="settings-menu__content show">
                <button class="settings-menu__button">
                  <span class="settings-menu__button-label">
                    Subtitles
                  </span>
                  <span class="settings-menu__button-value">
                    Off
                  </span>
                  <span class="settings-menu__arrow settings-menu__arrow--right">
                    <span class="svg">
                      <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                        <path d="M12.4 26l-3.333-3.333 7.4-6.667-7.4-6.6 3.333-3.4 10.6 10-10.6 10z"></path>
                      </svg>
                    </span>
                  </span>
                </button>
                <button class="settings-menu__button">
                  <span class="settings-menu__button-label">
                    Playback rate
                  </span>
                  <span class="settings-menu__button-value">
                    1
                  </span>
                  <span class="settings-menu__arrow settings-menu__arrow--right">
                    <span class="svg">
                      <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                        <path d="M12.4 26l-3.333-3.333 7.4-6.667-7.4-6.6 3.333-3.4 10.6 10-10.6 10z"></path>
                      </svg>
                    </span>
                  </span>
                </button>
              </div>
            </div>
          </mock:shadow-root>
        </xm-settings-menu>
        <input autocomplete="off" class="controls__slider" max="0" min="0" part="slider" step="any" type="range" value="0">
        <div class="controls__toolbar">
          <button title="Play">
            <span class="svg">
              <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                <path d="M24.267 14.867l-4.867-2.867c-1.067-0.6-2.867-1.667-3.933-2.333l-5.067-2.867c-1.067-0.667-2.133-0.133-2.133 1.133v16.067c0 1.267 1.067 1.8 2.133 1.133l4.933-2.867c1.067-0.6 2.933-1.667 4-2.333l4.867-2.867c1.133-0.533 1.133-1.6 0.067-2.2z"></path>
              </svg>
            </span>
          </button>
          <div class="controls-volume">
            <button class="controls__mute" title="Mute">
              <span class="svg">
                <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                  <path d="M16.754 4.159c-0.553 0-1.061 0.231-1.522 0.692l-5.211 4.473v13.373l5.211 4.565c0.277 0.277 0.599 0.461 0.968 0.507 0.277 0.046 0.507 0.046 0.784 0.046 0.277-0.046 0.507-0.138 0.692-0.277 0.461-0.277 0.784-0.738 0.784-1.337v-20.521c0-0.876-0.83-1.522-1.706-1.522v0zM8.315 9.323v13.373h-3.735c-0.646 0-1.199-0.184-1.614-0.599s-0.553-0.968-0.553-1.614v-9.038c0-0.646 0.184-1.199 0.646-1.614 0.461-0.369 1.061-0.553 1.752-0.553l3.505 0.046zM24.040 7.294c-0.692 0-1.291 0.599-1.291 1.291 0 0.461 0.277 0.876 0.646 1.107 1.107 0.692 2.029 1.66 2.675 2.813s0.968 2.444 0.968 3.735c0 1.291-0.323 2.582-0.968 3.735-0.599 1.153-1.476 2.167-2.628 2.813 0 0 0 0 0 0-0.415 0.277-0.646 0.646-0.738 1.015-0.092 0.323 0.046 0.692 0.231 0.922 0.369 0.507 1.199 0.692 1.891 0.231 1.43-0.922 2.675-2.213 3.505-3.735s1.245-3.228 1.245-4.98c0-1.752-0.415-3.459-1.291-4.98-0.83-1.522-2.075-2.859-3.551-3.781-0.138-0.138-0.415-0.184-0.692-0.184v0zM20.766 10.568c-0.692 0.138-1.199 0.83-1.061 1.522 0.046 0.369 0.323 0.692 0.646 0.876 0.553 0.323 1.015 0.784 1.337 1.337s0.507 1.199 0.507 1.845c0 0.646-0.138 1.337-0.461 1.891-0.323 0.599-0.784 1.107-1.383 1.43 0 0 0 0 0 0-0.415 0.277-0.646 0.599-0.692 0.922-0.092 0.323 0 0.646 0.184 0.922 0.369 0.507 1.107 0.784 1.845 0.369 0 0 0 0 0 0 0.968-0.553 1.752-1.43 2.306-2.398s0.83-2.075 0.784-3.228c-0.046-1.107-0.323-2.167-0.876-3.090s-1.337-1.706-2.26-2.213c-0.277-0.138-0.553-0.231-0.876-0.184v0z"></path>
                </svg>
              </span>
            </button>
            <input class="controls__slider controls__slider--volume" max="1" min="0" step="0.1" type="range" value="1">
          </div>
          <span class="controls__time" title="Time left of 00:00s">
            00:00 / 00:00
          </span>
          <div class="controls__playback-rate">
            <div class="controls__playback-rate__menu">
              <button class="controls__playback-rate__button"></button>
              <button class="controls__playback-rate__button"></button>
              <button class="controls__playback-rate__button controls__playback-rate__button--current"></button>
              <button class="controls__playback-rate__button"></button>
              <button class="controls__playback-rate__button"></button>
              <button class="controls__playback-rate__button"></button>
              <button class="controls__playback-rate__button"></button>
            </div>
            <button title="Playback rate">
              <span class="controls__shortcut-icon svg">
                <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                  <path d="M6.857 20.571q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM10.286 12.571q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM17.929 21.161l1.804-6.821q0.107-0.464-0.134-0.866t-0.688-0.527-0.857 0.116-0.536 0.705l-1.804 6.821q-1.071 0.089-1.911 0.777t-1.125 1.759q-0.357 1.375 0.357 2.607t2.089 1.589 2.607-0.357 1.589-2.089q0.286-1.071-0.107-2.089t-1.286-1.625zM29.714 20.571q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM18.286 9.143q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM26.286 12.571q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM32 20.571q0 4.661-2.518 8.625-0.339 0.518-0.964 0.518h-25.036q-0.625 0-0.964-0.518-2.518-3.946-2.518-8.625 0-3.25 1.268-6.214t3.411-5.107 5.107-3.411 6.214-1.268 6.214 1.268 5.107 3.411 3.411 5.107 1.268 6.214z"></path>
                </svg>
              </span>
            </button>
          </div>
          <button title="Settings">
            <span class="controls__settings-icon controls__settings-icon--open svg">
              <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                <path d="M18.286 16q0-1.893-1.339-3.232t-3.232-1.339-3.232 1.339-1.339 3.232 1.339 3.232 3.232 1.339 3.232-1.339 1.339-3.232zM27.429 14.054v3.964q0 0.214-0.143 0.411t-0.357 0.232l-3.304 0.5q-0.339 0.964-0.696 1.625 0.625 0.893 1.911 2.464 0.179 0.214 0.179 0.446t-0.161 0.411q-0.482 0.661-1.768 1.929t-1.679 1.268q-0.214 0-0.464-0.161l-2.464-1.929q-0.786 0.411-1.625 0.679-0.286 2.429-0.518 3.321-0.125 0.5-0.643 0.5h-3.964q-0.25 0-0.438-0.152t-0.205-0.384l-0.5-3.286q-0.875-0.286-1.607-0.661l-2.518 1.911q-0.179 0.161-0.446 0.161-0.25 0-0.446-0.196-2.25-2.036-2.946-3-0.125-0.179-0.125-0.411 0-0.214 0.143-0.411 0.268-0.375 0.911-1.188t0.964-1.259q-0.482-0.893-0.732-1.768l-3.268-0.482q-0.232-0.036-0.375-0.223t-0.143-0.42v-3.964q0-0.214 0.143-0.411t0.339-0.232l3.321-0.5q0.25-0.821 0.696-1.643-0.714-1.018-1.911-2.464-0.179-0.214-0.179-0.429 0-0.179 0.161-0.411 0.464-0.643 1.759-1.92t1.688-1.277q0.232 0 0.464 0.179l2.464 1.911q0.786-0.411 1.625-0.679 0.286-2.429 0.518-3.321 0.125-0.5 0.643-0.5h3.964q0.25 0 0.438 0.152t0.205 0.384l0.5 3.286q0.875 0.286 1.607 0.661l2.536-1.911q0.161-0.161 0.429-0.161 0.232 0 0.446 0.179 2.304 2.125 2.946 3.036 0.125 0.143 0.125 0.393 0 0.214-0.143 0.411-0.268 0.375-0.911 1.188t-0.964 1.259q0.464 0.893 0.732 1.75l3.268 0.5q0.232 0.036 0.375 0.223t0.143 0.42z"></path>
              </svg>
            </span>
          </button>
          <button title="Enter full screen">
            <span class="svg">
              <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                <path d="M10.8 18.867l-5 5-3-2.933v8.267h8.267l-2.933-3 5-5zM20.933 2.8l2.933 3-5 5 2.333 2.333 5-5 3 2.933v-8.267zM10.8 13.133l2.333-2.333-5-5 2.933-3h-8.267v8.267l3-2.933zM21.2 18.867l-2.333 2.333 5 5-2.933 3h8.267v-8.267l-3 2.933z"></path>
              </svg>
            </span>
          </button>
        </div>
      </div>
    </mock:shadow-root>
  </xm-controls>
</div>
`;

exports[`xm-controls should show a list of texttracks 1`] = `
<div>
  <xm-controls>
    <mock:shadow-root>
      <div class="controls">
        <xm-settings-menu>
          <mock:shadow-root>
            <div class="menu settings-menu">
              <div class="menu menu--open settings-menu__submenu">
                <div class="settings-menu__submenu-header">
                  <button class="settings-menu__button">
                    <span class="settings-menu__arrow settings-menu__arrow--left">
                      <span class="svg">
                        <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                          <path d="M19.6 6l3.333 3.333-7.4 6.667 7.4 6.6-3.333 3.4-10.533-10 10.533-10z"></path>
                        </svg>
                      </span>
                    </span>
                    <span class="settings-menu__button-label">
                      Subtitles
                    </span>
                  </button>
                </div>
                <div class="settings-menu__submenu-content">
                  <button class="settings-menu__button">
                    Off
                  </button>
                  <button class="settings-menu__button settings-menu__button--current">
                    Deutsch
                  </button>
                </div>
              </div>
              <div class="hide settings-menu__content">
                <button class="settings-menu__button">
                  <span class="settings-menu__button-label">
                    Subtitles
                  </span>
                  <span class="settings-menu__button-value">
                    Deutsch
                  </span>
                  <span class="settings-menu__arrow settings-menu__arrow--right">
                    <span class="svg">
                      <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                        <path d="M12.4 26l-3.333-3.333 7.4-6.667-7.4-6.6 3.333-3.4 10.6 10-10.6 10z"></path>
                      </svg>
                    </span>
                  </span>
                </button>
                <button class="settings-menu__button">
                  <span class="settings-menu__button-label">
                    Playback rate
                  </span>
                  <span class="settings-menu__button-value">
                    1
                  </span>
                  <span class="settings-menu__arrow settings-menu__arrow--right">
                    <span class="svg">
                      <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                        <path d="M12.4 26l-3.333-3.333 7.4-6.667-7.4-6.6 3.333-3.4 10.6 10-10.6 10z"></path>
                      </svg>
                    </span>
                  </span>
                </button>
              </div>
            </div>
          </mock:shadow-root>
        </xm-settings-menu>
        <input autocomplete="off" class="controls__slider" max="0" min="0" part="slider" step="any" type="range" value="0">
        <div class="controls__toolbar">
          <button title="Play">
            <span class="svg">
              <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                <path d="M24.267 14.867l-4.867-2.867c-1.067-0.6-2.867-1.667-3.933-2.333l-5.067-2.867c-1.067-0.667-2.133-0.133-2.133 1.133v16.067c0 1.267 1.067 1.8 2.133 1.133l4.933-2.867c1.067-0.6 2.933-1.667 4-2.333l4.867-2.867c1.133-0.533 1.133-1.6 0.067-2.2z"></path>
              </svg>
            </span>
          </button>
          <div class="controls-volume">
            <button class="controls__mute" title="Mute">
              <span class="svg">
                <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                  <path d="M16.754 4.159c-0.553 0-1.061 0.231-1.522 0.692l-5.211 4.473v13.373l5.211 4.565c0.277 0.277 0.599 0.461 0.968 0.507 0.277 0.046 0.507 0.046 0.784 0.046 0.277-0.046 0.507-0.138 0.692-0.277 0.461-0.277 0.784-0.738 0.784-1.337v-20.521c0-0.876-0.83-1.522-1.706-1.522v0zM8.315 9.323v13.373h-3.735c-0.646 0-1.199-0.184-1.614-0.599s-0.553-0.968-0.553-1.614v-9.038c0-0.646 0.184-1.199 0.646-1.614 0.461-0.369 1.061-0.553 1.752-0.553l3.505 0.046zM24.040 7.294c-0.692 0-1.291 0.599-1.291 1.291 0 0.461 0.277 0.876 0.646 1.107 1.107 0.692 2.029 1.66 2.675 2.813s0.968 2.444 0.968 3.735c0 1.291-0.323 2.582-0.968 3.735-0.599 1.153-1.476 2.167-2.628 2.813 0 0 0 0 0 0-0.415 0.277-0.646 0.646-0.738 1.015-0.092 0.323 0.046 0.692 0.231 0.922 0.369 0.507 1.199 0.692 1.891 0.231 1.43-0.922 2.675-2.213 3.505-3.735s1.245-3.228 1.245-4.98c0-1.752-0.415-3.459-1.291-4.98-0.83-1.522-2.075-2.859-3.551-3.781-0.138-0.138-0.415-0.184-0.692-0.184v0zM20.766 10.568c-0.692 0.138-1.199 0.83-1.061 1.522 0.046 0.369 0.323 0.692 0.646 0.876 0.553 0.323 1.015 0.784 1.337 1.337s0.507 1.199 0.507 1.845c0 0.646-0.138 1.337-0.461 1.891-0.323 0.599-0.784 1.107-1.383 1.43 0 0 0 0 0 0-0.415 0.277-0.646 0.599-0.692 0.922-0.092 0.323 0 0.646 0.184 0.922 0.369 0.507 1.107 0.784 1.845 0.369 0 0 0 0 0 0 0.968-0.553 1.752-1.43 2.306-2.398s0.83-2.075 0.784-3.228c-0.046-1.107-0.323-2.167-0.876-3.090s-1.337-1.706-2.26-2.213c-0.277-0.138-0.553-0.231-0.876-0.184v0z"></path>
                </svg>
              </span>
            </button>
            <input class="controls__slider controls__slider--volume" max="1" min="0" step="0.1" type="range" value="1">
          </div>
          <span class="controls__time" title="Time left of 00:00s">
            00:00 / 00:00
          </span>
          <div class="controls__playback-rate">
            <div class="controls__playback-rate__menu">
              <button class="controls__playback-rate__button"></button>
              <button class="controls__playback-rate__button"></button>
              <button class="controls__playback-rate__button controls__playback-rate__button--current"></button>
              <button class="controls__playback-rate__button"></button>
              <button class="controls__playback-rate__button"></button>
              <button class="controls__playback-rate__button"></button>
              <button class="controls__playback-rate__button"></button>
            </div>
            <button title="Playback rate">
              <span class="controls__shortcut-icon svg">
                <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                  <path d="M6.857 20.571q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM10.286 12.571q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM17.929 21.161l1.804-6.821q0.107-0.464-0.134-0.866t-0.688-0.527-0.857 0.116-0.536 0.705l-1.804 6.821q-1.071 0.089-1.911 0.777t-1.125 1.759q-0.357 1.375 0.357 2.607t2.089 1.589 2.607-0.357 1.589-2.089q0.286-1.071-0.107-2.089t-1.286-1.625zM29.714 20.571q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM18.286 9.143q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM26.286 12.571q0-0.946-0.67-1.616t-1.616-0.67-1.616 0.67-0.67 1.616 0.67 1.616 1.616 0.67 1.616-0.67 0.67-1.616zM32 20.571q0 4.661-2.518 8.625-0.339 0.518-0.964 0.518h-25.036q-0.625 0-0.964-0.518-2.518-3.946-2.518-8.625 0-3.25 1.268-6.214t3.411-5.107 5.107-3.411 6.214-1.268 6.214 1.268 5.107 3.411 3.411 5.107 1.268 6.214z"></path>
                </svg>
              </span>
            </button>
          </div>
          <button class="controls__subtitle-button" title="Disable subtitles">
            <span class="controls__shortcut-icon controls__shortcut-icon--active svg">
              <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                <path d="M29 4h-26c-1.656 0-3 1.344-3 3v18c0 1.656 1.344 3 3 3h26c1.656 0 3-1.344 3-3v-18c0-1.656-1.344-3-3-3zM13.631 17.981c0.175-0.156 0.444-0.131 0.575 0.056l1.219 1.731c0.106 0.15 0.094 0.35-0.031 0.481-3.35 3.55-10.8 2.006-10.8-4.244 0-6.081 7.606-7.469 10.781-4.381 0.131 0.125 0.156 0.2 0.063 0.356l-1.094 1.906c-0.119 0.194-0.387 0.25-0.569 0.106-2.55-2-5.912-0.931-5.912 1.95 0.006 3 3.194 4.406 5.769 2.037zM25.531 17.981c0.175-0.156 0.444-0.131 0.575 0.056l1.219 1.731c0.106 0.15 0.094 0.35-0.031 0.481-3.344 3.556-10.794 2.006-10.794-4.244 0-6.081 7.606-7.469 10.781-4.381 0.131 0.125 0.156 0.2 0.063 0.356l-1.094 1.906c-0.119 0.194-0.387 0.25-0.569 0.106-2.55-2-5.913-0.931-5.913 1.95 0 3 3.188 4.406 5.762 2.037z"></path>
              </svg>
            </span>
          </button>
          <button title="Settings">
            <span class="controls__settings-icon svg">
              <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                <path d="M18.286 16q0-1.893-1.339-3.232t-3.232-1.339-3.232 1.339-1.339 3.232 1.339 3.232 3.232 1.339 3.232-1.339 1.339-3.232zM27.429 14.054v3.964q0 0.214-0.143 0.411t-0.357 0.232l-3.304 0.5q-0.339 0.964-0.696 1.625 0.625 0.893 1.911 2.464 0.179 0.214 0.179 0.446t-0.161 0.411q-0.482 0.661-1.768 1.929t-1.679 1.268q-0.214 0-0.464-0.161l-2.464-1.929q-0.786 0.411-1.625 0.679-0.286 2.429-0.518 3.321-0.125 0.5-0.643 0.5h-3.964q-0.25 0-0.438-0.152t-0.205-0.384l-0.5-3.286q-0.875-0.286-1.607-0.661l-2.518 1.911q-0.179 0.161-0.446 0.161-0.25 0-0.446-0.196-2.25-2.036-2.946-3-0.125-0.179-0.125-0.411 0-0.214 0.143-0.411 0.268-0.375 0.911-1.188t0.964-1.259q-0.482-0.893-0.732-1.768l-3.268-0.482q-0.232-0.036-0.375-0.223t-0.143-0.42v-3.964q0-0.214 0.143-0.411t0.339-0.232l3.321-0.5q0.25-0.821 0.696-1.643-0.714-1.018-1.911-2.464-0.179-0.214-0.179-0.429 0-0.179 0.161-0.411 0.464-0.643 1.759-1.92t1.688-1.277q0.232 0 0.464 0.179l2.464 1.911q0.786-0.411 1.625-0.679 0.286-2.429 0.518-3.321 0.125-0.5 0.643-0.5h3.964q0.25 0 0.438 0.152t0.205 0.384l0.5 3.286q0.875 0.286 1.607 0.661l2.536-1.911q0.161-0.161 0.429-0.161 0.232 0 0.446 0.179 2.304 2.125 2.946 3.036 0.125 0.143 0.125 0.393 0 0.214-0.143 0.411-0.268 0.375-0.911 1.188t-0.964 1.259q0.464 0.893 0.732 1.75l3.268 0.5q0.232 0.036 0.375 0.223t0.143 0.42z"></path>
              </svg>
            </span>
          </button>
          <button title="Enter full screen">
            <span class="svg">
              <svg focusable="false" role="presentation" viewBox="0 0 32 32">
                <path d="M10.8 18.867l-5 5-3-2.933v8.267h8.267l-2.933-3 5-5zM20.933 2.8l2.933 3-5 5 2.333 2.333 5-5 3 2.933v-8.267zM10.8 13.133l2.333-2.333-5-5 2.933-3h-8.267v8.267l3-2.933zM21.2 18.867l-2.333 2.333 5 5-2.933 3h8.267v-8.267l-3 2.933z"></path>
              </svg>
            </span>
          </button>
        </div>
      </div>
    </mock:shadow-root>
  </xm-controls>
</div>
`;
